pkg/daemon/controller/controller.go

```diff
@@ -202,6 +202,128 @@ func (sc *ServiceController) Reconcile(_ context.Context, req ctrl.Request) (res
 	return ctrl.Result{}, nil
 }
 
+// Cleanup removes an instance from the cluster. If the parameter flag is true, the node object is not deleted.
+func (sc *ServiceController) Cleanup(preserveNode bool) error {
+	var node core.Node
+	err := sc.client.Get(sc.ctx, client.ObjectKey{Name: sc.nodeName}, &node)
+	if err != nil {
+		return err
+	}
+
+	desiredVersion, present := node.Annotations[nodeconfig.DesiredVersionAnnotation]
+	if !present {
+		return errors.Errorf("node missing desired version annotation")
+	}
+	// Fetch the CM of the desired version
+	cm := &core.ConfigMap{}
+	err = sc.client.Get(sc.ctx,
+		client.ObjectKey{Namespace: sc.watchNamespace, Name: servicescm.NamePrefix + desiredVersion}, cm)
+	if err != nil {
+		return err
+	}
+	return sc.removeServices(cm, node, preserveNode)
+}
+
+// removeServices removes all the services defined in the given ConfigMap from this instance.
+// The given Node object is deleted if the given flag is false
+func (sc *ServiceController) removeServices(cm *core.ConfigMap, node core.Node, preserveNode bool) error {
+	klog.Infof("removing services defined in ConfigMap %s", cm.Name)
+	cmData, err := servicescm.Parse(cm.Data)
+	if err != nil {
+		return err
+	}
+	for _, service := range cmData.Services {
+		if err := sc.ensureServiceIsRemoved(service.Name); err != nil {
+			return err
```

Comment:
We should collect the errors and not fail on the first service we were not able to remove.